src/shop.c

```diff
@@ -0,0 +1,1368 @@
+static void Task_BuyHowManyDialogueInit(u8 taskId);
+static void Task_BuyHowManyDialogueHandleInput(u8 taskId);
+static void CreateBuyMenuConfirmPurchaseWindow(u8 taskId);
+static void BuyMenuTryMakePurchase(u8 taskId);
+static void BuyMenuSubtractMoney(u8 taskId);
+static void Task_ReturnToItemListAfterItemPurchase(u8 taskId);
+static void BuyMenuReturnToItemList(u8 taskId);
+static void ExitBuyMenu(u8 taskId);
+static void Task_ExitBuyMenu(u8 taskId);
+static void nullsub_52(u8 taskId);
+static void nullsub_53(void);
+static void RecordQuestLogItemPurchase(void);
+
+extern u8 MapGridGetMetatileLayerTypeAt(s16 x, s16 y);
+extern u16 BagGetQuantityByItemId(u16 item);
+
+static const struct MenuAction sShopMenuActions_BuySellQuit[] =
+{
+    {gText_ShopBuy, {.void_u8 = Task_HandleShopMenuBuy}},
+    {gText_ShopSell, {.void_u8 = Task_HandleShopMenuSell}},
+    {gText_ShopQuit, {.void_u8 = Task_HandleShopMenuQuit}}
+};
+
+static const struct YesNoFuncTable sShopMenuActions_BuyQuit[] =
+{
+    BuyMenuTryMakePurchase,
+    BuyMenuReturnToItemList
+};
+
+static const struct WindowTemplate sShopMenuWindowTemplates[] =
+{
+    {
+        .bg = 0,
+        .tilemapLeft = 2,
+        .tilemapTop = 1,
+        .width = 12,
+        .height = 6,
+        .paletteNum = 15,
+        .baseBlock = 8
+    }
+};    
+
+static const struct BgTemplate sShopBuyMenuBgTemplates[4] =
+{
+    {
+        .bg = 0,
+        .charBaseIndex = 2,
+        .mapBaseIndex = 31,
+        .screenSize = 0,
+        .paletteMode = 0,
+        .priority = 0,
+        .baseTile = 0
+    },
+    {
+        .bg = 1,
+        .charBaseIndex = 0,
+        .mapBaseIndex = 30,
+        .screenSize = 0,
+        .paletteMode = 0,
+        .priority = 1,
+        .baseTile = 0
+    },
+    {
+        .bg = 2,
+        .charBaseIndex = 0,
+        .mapBaseIndex = 29,
+        .screenSize = 0,
+        .paletteMode = 0,
+        .priority = 2,
+        .baseTile = 0
+    },
+    {
+        .bg = 3,
+        .charBaseIndex = 0,
+        .mapBaseIndex = 28,
+        .screenSize = 0,
+        .paletteMode = 0,
+        .priority = 3,
+        .baseTile = 0
+    }
+};
+
+// Functions
+static u8 CreateShopMenu(u8 a0)
+{
+    gShopData.martType = GetMartTypeFromItemList(a0) & 0xF;
+    gShopData.selectedRow = 0;
+    if (!(ContextNpcGetTextColor()))
+        gShopData.unk16_4 = 4;
+    else
+        gShopData.unk16_4 = 5;
+    
+    sShopMenuWindowId = AddWindow(sShopMenuWindowTemplates);
+    SetStdWindowBorderStyle(sShopMenuWindowId, 0);
+    PrintTextArray(sShopMenuWindowId, 2, GetMenuCursorDimensionByFont(2, 0), 2, 16, 3, sShopMenuActions_BuySellQuit);
+    Menu_InitCursor(sShopMenuWindowId, 2, 0, 2, 16, 3, 0);
+    PutWindowTilemap(sShopMenuWindowId);
+    CopyWindowToVram(sShopMenuWindowId, 1);
+    return CreateTask(Task_ShopMenu, 8);
+}
+
+static u8 GetMartTypeFromItemList(u32 a0)
+{    
+    u16 i;
+    
+    if (a0)
+        return (u8)a0;
+    goto MAIN;
```

Comment:
Use goto only when you really have no idea. If so, please setup compiler explorer for this

Reply:
there's a `mov r0, #1` and return statement at the top of the function that is jumped to later - I wasn't sure if there was a better way. But here's the compiler explorer: https://cexplore.karathan.at/z/9jH79p

Reply:
https://cexplore.karathan.at/z/Eln5Lf